source/geod24/concurrency.d

```diff
@@ -1,2583 +1,1608 @@
 
-    private:
-        void switchContext() nothrow
-        {
-            if (mutex_nothrow) mutex_nothrow.unlock_nothrow();
-            scope (exit)
-                if (mutex_nothrow)
-                    mutex_nothrow.lock_nothrow();
-            this.outer.yield();
-        }
+        Iteratively joins all tracked threads.  This function will block add,
+        remove, and opApply until it completes.
 
-        private bool notified;
-    }
+        Params:
+            rethrow = Rethrow any unhandled exception which may have caused the
+                      current thread to terminate.
 
-private:
-    void dispatch()
-    {
-        import std.algorithm.mutation : remove;
+        Throws:
+            Any exception not handled by the joined threads.
 
-        while (m_fibers.length > 0)
+    ***************************************************************************/
+
+    final void joinAll ( bool rethrow = true )
+    {
+        synchronized( this )
         {
-            auto t = m_fibers[m_pos].call(Fiber.Rethrow.no);
-            if (t !is null && !(cast(OwnerTerminated) t))
-            {
-                throw t;
-            }
-            if (m_fibers[m_pos].state == Fiber.State.TERM)
-            {
-                if (m_pos >= (m_fibers = remove(m_fibers, m_pos)).length)
-                    m_pos = 0;
-            }
-            else if (m_pos++ >= m_fibers.length - 1)
+            // NOTE: This loop relies on the knowledge that m_all uses the
+            //       Thread object for both the key and the mapped value.
+            foreach ( Thread t; m_all.keys )
             {
-                m_pos = 0;
+                t.join( rethrow );
             }
         }
     }
 
-private:
-    Fiber[] m_fibers;
-    size_t m_pos;
 }
 
-@system unittest
+
+/*******************************************************************************
+
+    Iteratively joins all tracked threads.  This function will block add,
+    remove, and opApply until it completes.
+
+*******************************************************************************/
+
+public void joinAllThread ()
+{
+    ThreadScheduler.instance.joinAll();
+}
+
+
+/*******************************************************************************
+
+    Clean all `InfoObjects` in use from main thread.
+
+*******************************************************************************/
+
+public void cleanupMainThread ()
+{
+    thisInfo.cleanup(true);
+}
+
+
+/*******************************************************************************
+
+    An example Scheduler using Fibers.
+
+    This is an example scheduler that creates a new Fiber per call to spawn
+    and multiplexes the execution of all fibers within the main thread.
+
+*******************************************************************************/
+
+class FiberScheduler : Scheduler, InfoObject
 {
-    static void receive(Condition cond, ref size_t received)
+    private shared(bool) terminated;
```

Comment:
Why are `terminated` and `terminated_time` marked as `shared`? They are private and only used within this class. Also, `terminated_time` is not read anywhere, it is only written to inside of `stop()`, but never used.